### ISSUE TITLE
Viewporter protocol and layer-shell surfaces

### ISSUE THREAD

**Issue by Tamaranch** [labwc/labwc#2129]
Hi, I'm currently trying, in xfce4-panel, to replace autohide by layer-shell surface displacement (using negative margins, a bit like on X11) by using the viewporter protocol instead, as I've been advised several times here and there.

This works fine if the panel is anchored at the top of the screen, but if it's anchored at the bottom, I'm faced with the following problem: cropping does remove the top of the panel, but without preserving the anchoring, which gives the impression that the panel is hiding upwards instead of downwards.

To illustrate, here are a few screenshots:

Full panel: ![panel-full](https://github.com/user-attachments/assets/93ebc839-8675-4b98-922c-f564425dc140)

Panel anchored at the top and half-hidden: ![panel-top](https://github.com/user-attachments/assets/8c7466fd-c478-44f9-ae9d-532f3bf4291a)

Panel anchored at bottom and half-hidden: ![panel-bottom](https://github.com/user-attachments/assets/6c7b5971-c741-4375-832b-be21d5592749)

Do you think it would be possible to preserve anchoring when using this protocol for layer-shell surfaces?

Note that this behavior is not specific to labwc: at least sway and wayfire behave in the same way.

**Comment by johanmalm** [comment]
Hi. Thanks for writing. Hope you're well.

I'd not thought about using view-porter in that context. Could you send a link to the patch/PR/MR/tree?
Does the "usable area" (that windows can use) adjust when you hide the panel like that.

I had quietly thought we could try `wlr_scene_subsurface_tree_set_clip()` (from the compositor side) when using negative margins to avoid the surface showing outside the output. Or possibly `wlr_scene_buffer_set_source_box()`. But have to admit, have not been playing with it at all.

**Comment by droc12345** [comment]
I'm interested, as I was looking at viewporter to handle things like menus, being too long, ie scrolling off screen.

**Comment by Tamaranch** [comment]
> Hi. Thanks for writing. Hope you're well.

Yes, I'm fine thanks! :)

> Could you send a link to the patch/PR/MR/tree?

I haven't put anything online yet, I just have a local draft. But, apart from the wayland boilerplate, to get the screenshots above, it boils down to replacing
```c
gtk_layer_set_margin (window, GTK_LAYER_SHELL_EDGE_TOP, -(window_height / 2));
```
with
```c
wp_viewport_set_source (wp_viewport, wl_fixed_from_int (0), wl_fixed_from_int (0), wl_fixed_from_int (window_width), wl_fixed_from_int (window_height / 2));
```
when the panel is anchored at the top of the screen, and
```c
gtk_layer_set_margin (window, GTK_LAYER_SHELL_EDGE_BOTTOM, -(window_height / 2));
```
with
```c
wp_viewport_set_source (wp_viewport, wl_fixed_from_int (0), wl_fixed_from_int (window_height / 2), wl_fixed_from_int (window_width), wl_fixed_from_int (window_height / 2));
```
when it's anchored at the bottom.

> Does the "usable area" (that windows can use) adjust when you hide the panel like that.

You're referring to the layer-shell protocol exclusive zone, I guess? We disable it when autohide is enabled, so that the panel temporarily covers the windows when it shows up. But I've just tested to see (with the panel anchored at the top of the screen) and no, the exclusive zone remains unchanged (which seems a bit odd too).

**Comment by Consolatis** [comment]
Yeah, that looks like a bug. Not sure where; if sway and especially wayfire (which doesn't use wlr_scene) behave the same it might be wlroots itself. Or all 3 cargo culted the same bug along (which is absolutely possible).

@vyivel any idea?

**Comment by vyivel** [comment]
I'd say this is a compositor(s) bug.

Is there a xfce4-panel branch I can test with?

**Comment by Tamaranch** [comment]
> Is there a xfce4-panel branch I can test with?

No, I've only got a local draft, I don't think I'll have time to push something even remotely clean in the near future, sorry.

**Comment by vyivel** [comment]
Sway/Wayfire/labwc don't position layer surfaces correctly if their size doesn't match the size sent with `zwlr_layer_surface_v1.configure`. (Neither does `wlr_scene_layer_surface_v1_configure()`.)

If you're interested in fixing this, I have a simple script to help with debugging: https://codeberg.org/vyivel/layercheck/src/branch/main/examples/bottom-half.sh

To the best of my understanding, the correct behavior would be in the bottom left.

![image](https://github.com/user-attachments/assets/101132ed-6fb0-4945-84c4-13802b19621f)

**Comment by Consolatis** [comment]
I've added the viewporter protocol to your example: https://codeberg.org/vyivel/layercheck/pulls/1
That PR shows that `zwlr_layer_surface_v1_set_size()` indeed makes the approach work but I am still not sure what is the correct way for a compositor to calculate the anchor positions as we have 3 different sizes:
- layershell size
- buffer size
- viewporter size
